### ISSUE TITLE
update ntpd.conf generation to support pool directive for public pool servers

### ISSUE THREAD

**Issue by kevinfason** [opnsense/core#5569]
**Important notices**

Before you add a new report, we ask you kindly to acknowledge the following:

- [X ] I have read the contributing guide lines at https://github.com/opnsense/core/blob/master/CONTRIBUTING.md
- [X ] I am convinced that my issue is new after having checked both open and closed issues at https://github.com/opnsense/core/issues?q=is%3Aissue

**Is your feature request related to a problem? Please describe.**

ntpd.conf that is generated for newer ntpd daemons that supports the pool directive.

A clear and concise description of what the problem is including your motivation for the request,
i.e. "For the purpose of [...] I am missing a solution that will [...]."

**Describe the solution you like**

When ntpd.conf is generated and it includes servers from the public pool (*.pool.ntp.org) use the _pool_ directive instead of the _server_ directive. This tells (4.2.6+) ntpd to act upon that Time Server differently.

This directive facilitates features such as:

- Removal of duplicate servers
- DNS lookups for servers are done at ntpd startup whereas pool will perform a new DNS lookup if the current host no longer responds. Useful for long uptime devices such as OPNSense.
- Ability to use a single pool directive line (don't know how this is impacted by the vendor *.opnsense.pool.ntp.org maybe can just use opnsense.pool.ntp.org now vs 0-4.opnsense.pool.ntp.org?)

Therefore, I would propose that any Time servers entered into Services | Network Time | General that is from the pool such as the default *.opnsense.pool.ntp.org entires, or user-entered is parsed and if it is *.pool.ntp.org host it is called out with the pool directive instead of the server directive so ntpd treats them appropriately.

**Describe alternatives you considered**

I am using 21.7.8 currently. I tried to modify my /var/etc/ntpd.conf to use pool for 0.opnsense.pool.ntp.org and restart the service but restarting via the GUI and via cli (pluginctl -s ntp restart) regenerates the conf file.

**Additional context**

More detail:

- http://doc.ntp.org/archives/4.2.8-series/discover/#server-pool-scheme
- https://kb.meinbergglobal.com/kb/time_sync/ntp/configuration/ntp_pool_usage
- (Comment by rlaager) https://www.reddit.com/r/linuxadmin/comments/lhnlnv/configuring_etcntpconf_best_practices/

Looking at my ntpd.conf in 21.7.8 there is the 'Upstream Servers' comment, maybe put the pool entries under their own so they are grouped together? 

Related however might be too picky here but I'm happy to write something up for submittal. On the help for Time servers mention something about not mixing pool entries such as the opnsense vendor and the regional ones? Mention using multiple sources if not using pool, ie 4-7 recommended? Perhaps mention something about not mixing smearing vs leap second sources like Google?

**Comment by AdSchellevis** [comment]
If you want to experiment a bit with different options, the upstream servers are merged here : https://github.com/opnsense/core/blob/df7882cfa961532f058bd3790775184be3041b54/src/etc/inc/plugins.inc.d/ntpd.inc#L327-L340

(/usr/local/etc/inc/plugins.inc.d/ntpd.inc)

`pluginctl` will generate the config based on these settings.

**Comment by kevinfason** [comment]
I had some time recently to mess with this and changed line 326 

`$ntpcfg .= "server {$ts}";`

to

```
/* Determine if Network Time Server is from the NTP Pool or not */
if (preg_match("/\.pool\.ntp\.org$/", $ts)) {
     $ntpcfg .= "pool {$ts}";
    } else {
     $ntpcfg .= "server {$ts}";
 }
```
This populates the relevant section of /var/etc/ntp.conf as expected

```
# Upstream Servers
server time.facebook.com maxpoll 9 prefer
server time.cloudflare.com maxpoll 9 prefer
pool 0.us.pool.ntp.org maxpoll 9 prefer
pool 1.us.pool.ntp.org maxpoll 9 prefer
```

I'm still testing but I have it on 3 systems currently.

**Comment by fichtner** [comment]
@kevinfason Looks reasonable, can you open PR?

**Comment by kevinfason** [comment]
Not quite ready yet as there is a functional piece as well as the configuration piece I already suggested so I am working on that part. I originally thought it was some FW rule but it behaves the same on fresh install. While the conf is setup correctly ntpd is not using them at all. 

ntpq peers is showing it not using them:

```
ntpq> peers
     remote           refid      st t when poll reach   delay   offset  jitter
==============================================================================
*GPS_NMEA(0)     .GPS.            0 l   10   16  377    0.000   +1.728   1.371
 **0.us.pool.ntp.o .POOL.          16 p    -   64    0    0.000   +0.000   0.000
 1.us.pool.ntp.o .POOL.          16 p    -   64    0    0.000   +0.000   0.000**
+time5.facebook. .FB...           1 u   45   64  377   12.962   +4.286   3.442
+time.cloudflare 10.98.8.212      3 u   45   64  377   13.281   +4.422   3.276
```

And ntpd logs are cycling through hosts
```
...Soliciting pool server 1.2.3.4
...Soliciting pool server 4.3.2.1
```

Turns out the 'nopeer' restriction is the problem and we have to generate another restriction to support the pool directive. Its spoken about [here](https://bugs.ntp.org/show_bug.cgi?id=2657) and man 5 ntp.conf states the following:

```
             nopeer  Deny unauthenticated packets which would result in
                     mobilizing a new association.  This includes broadcast
                     and symmetric active packets when a configured
                     association does not exist.  It also includes pool
                     associations, so if you want to use servers from a pool
                     directive and also want to use nopeer by default, you'll
                     want a restrict source ... line as well that does not
                     include the nopeer directive.
```
 Basically adding these two lines

```
restrict source limited kod nomodify notrap noquery
restrict -6 source limited kod nomodify notrap noquery
```

I'm working on that now and will do a PR. I'd like to test more and validate the restrict statements from documentation.

**Comment by kevinfason** [comment]
Think I got it all figured out. Pool directive works fine now once it has the above restrict line added. Note the one, not the two previously mentioned. The 'restrict source -6' for IPv6 support does not work and generates a syntax error in the logs so would need upstream to fix it. With that said I have the following at line 424 to enable pool functionality.  Above creates the conf file correctly and this lights up the functionality. 

```
        /* Access restrictions for pool */
    $ntpcfg .= "\nrestrict source";
    if (empty($config['ntpd']['kod'])) { /*note: this one works backwards */
        $ntpcfg .= ' kod';
    }
    if (empty($config['ntpd']['limited'])) { /*note: this one works backwards */
        $ntpcfg .= ' limited';
    }
    if (empty($config['ntpd']['nomodify'])) { /*note: this one works backwards */
        $ntpcfg .= ' nomodify';
    }
    if (!empty($config['ntpd']['noquery'])) {
        $ntpcfg .= ' noquery';
    }
    if (empty($config['ntpd']['notrap'])) { /*note: this one works backwards */
        $ntpcfg .= ' notrap';
    }
    if (!empty($config['ntpd']['noserve'])) {
        $ntpcfg .= ' noserve';
    }

    /* restrict source -6 does not work until upstream fixes */
```

I'll put in a PR here shortly as I want to add a function to the GUI as well that applies IMO. Its 'cos maxcount'. It tells NTPd how many servers max to maintain. default is 10. Some may want to adjust it one way or the other. Documentation states it should be odd number so falsetickers and outliers get addressed more easily. I was thinking 9 to be the default (see below) as it would put it in between 4-7 sources that documentation states is ideal.

![2023-02-15_20-39-08](https://user-images.githubusercontent.com/99461833/219263964-3709a290-7c0e-4340-9ba2-371cde95d5d9.png)

Reason being how it counts sources as it complicates matters since it subtracts the pool DNS servers from sources. With the pool directive, NTP will use more than one server IP from that DNS, such as 0.opnsense.pool.ntp.org. Four per pool entry if I recall. So with the default of 0-3.opnsense.pool.ntp.org it would start with 16 and settle down to 10 entries based on the sources declared however 6 would be sources as the four default entries would be subtracted. In this screenshot I have 0/1.opnsense.ntp.org pool ones and time.cloudflare.com, time.facebook.com declared along with my serial GPS so its polling 8. Ten default maxcount minus two pool ones. You can use say us.pool.ntp.org only and it will populate 10 sources from it and maintain them.

The maxcount in the GUI can mirror the orphan function already present which I think I can do up in that code so will give it a try to test out. 

Hopefully I explained this well.
